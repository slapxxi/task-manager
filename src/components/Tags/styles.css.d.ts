export const tags: string;
export const tag: string;
export const label: string;
export const input: string;
export const checkbox: string;
