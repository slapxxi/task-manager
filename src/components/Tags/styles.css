.tags {
  display: flex;
  align-items: center;
  flex-wrap: wrap;
  margin: 0;
  padding: 0;
  list-style: none;
}

.tag {
  overflow: hidden;
  box-sizing: border-box;
  margin: 0 0.5em 0 0;
  padding: 5px 0;
}

.tag:last-child {
  margin-right: 0;
}

.tag:first-child {
  margin-left: 0;
}

.label {
  position: relative;
  box-sizing: border-box;
  padding: 5px 10px;
  min-width: 14px;
  border-radius: 5px;
  background: var(--light-green);
  color: #298561;
  text-align: center;
  font-size: 13px;
  cursor: pointer;
}

.input {
  flex: 1;
  min-width: 0;
  margin: 0;
  padding: 0;
  outline: 0;
  border: 0;
  font-size: 1em;

  -webkit-appearance: none;
}

input::placeholder {
  color: #99a;
}

.checkbox {
  position: absolute;
  opacity: 0;
  cursor: pointer;
}

.checkbox:checked ~ .label {
  background-color: #bad2f5;
  color: #5385cd;
}
