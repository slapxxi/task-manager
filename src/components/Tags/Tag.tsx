import { Tag as ITag } from '@local/types';
import * as React from 'react';
import styles from './styles.css';

interface Props {
  tag: ITag;
  selected?: boolean;
  onSelect?: (isSelected: boolean, tag: ITag) => void;
}

class Tag extends React.PureComponent<Props, {}> {
  public handleSelect = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (this.props.onSelect) {
      this.props.onSelect(
        e.currentTarget.checked ? true : false,
        this.props.tag,
      );
    }
  };

  public render() {
    const { tag, selected } = this.props;
    return (
      <>
        <input
          type="checkbox"
          className={styles.checkbox}
          id={tag.id}
          onChange={this.handleSelect}
          checked={selected}
          data-testid="checkbox"
        />
        <label htmlFor={tag.id} className={styles.label}>
          {tag.name}
        </label>
      </>
    );
  }
}

export default Tag;
