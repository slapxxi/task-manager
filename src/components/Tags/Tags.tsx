import { Tag as ITag } from '@local/types';
import * as React from 'react';
import styles from './styles.css';

interface RenderParams {
  tag: ITag;
}

interface Props {
  tags: ITag[];
  renderTag: (params: RenderParams) => React.ReactNode;
}

class Tags extends React.Component<Props, {}> {
  public render() {
    const { tags, renderTag } = this.props;
    return (
      <ul className={styles.tags} data-testid="tags">
        {tags.map((tag) => (
          <li className={styles.tag} key={tag.id}>
            {renderTag && renderTag({ tag })}
          </li>
        ))}
      </ul>
    );
  }
}

export default Tags;
