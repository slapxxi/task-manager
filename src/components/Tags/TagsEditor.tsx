import { Keys } from '@lib';
import { Tag as ITag, UserCreatedTag } from '@local/types';
import { includes, last } from 'lodash';
import * as React from 'react';
import styles from './styles.css';
import Tag from './Tag';
import Tags from './Tags';

interface Props {
  tags: ITag[];
  onAddTag?: (tag: UserCreatedTag) => void;
  onRemoveTag?: (tag: ITag) => void;
  onRemoveTags?: (tags: ITag[]) => void;
  className?: string;
}

interface State {
  selectedTags: ID[];
}

class TagsEditor extends React.Component<Props, State> {
  public state = { selectedTags: [] };

  public handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { onAddTag } = this.props;
    if (onAddTag) {
      const value = e.target.value;
      if (this.isValidTagName(value)) {
        onAddTag({ name: this.normalizeTagName(value) });
        e.target.value = '';
      }
    }
  };

  public handleSelect = (isSelected: boolean, tag: ITag) => {
    if (isSelected) {
      this.setState({ selectedTags: [...this.state.selectedTags, tag.id] });
    } else {
      this.setState({
        selectedTags: this.state.selectedTags.filter(
          (selectedID) => selectedID !== tag.id,
        ),
      });
    }
  };

  public handleKeyPress = (e: React.KeyboardEvent<HTMLInputElement>) => {
    const key = e.keyCode || e.charCode;
    if (key === Keys.enter) {
      // TODO test for various space chars
      if (e.currentTarget.value.trim() !== '') {
        if (this.props.onAddTag) {
          this.props.onAddTag({ name: e.currentTarget.value });
          this.resetInput(e);
          return;
        }
      }
    }
    if (key === Keys.backspace) {
      return this.handleRemove(e);
    }
  };

  public handleRemove = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (this.isInputEmpty(e) && this.props.tags.length > 0) {
      if (this.state.selectedTags.length > 1) {
        if (this.props.onRemoveTags) {
          const selectedTags = this.state.selectedTags;
          return this.setState({ selectedTags: [] }, () => {
            if (this.props.onRemoveTags) {
              return this.props.onRemoveTags(
                this.props.tags.filter((t) => includes(selectedTags, t.id)),
              );
            }
          });
        }
      }
    }
    if (this.props.onRemoveTag) {
      if (this.state.selectedTags.length === 1) {
        const selectedTag = this.state.selectedTags[0];
        return this.setState({ selectedTags: [] }, () =>
          this.props.onRemoveTag!(
            this.props.tags.filter((t) => t.id === selectedTag)[0],
          ),
        );
      }
      return this.props.onRemoveTag(last(this.props.tags)!);
    }
  };

  public render() {
    const { tags, className } = this.props;
    return (
      <div className={className}>
        <Tags
          tags={tags}
          renderTag={({ tag }) => (
            <Tag
              tag={tag}
              selected={this.isTagSelected(tag)}
              onSelect={this.handleSelect}
            />
          )}
        />
        <input
          placeholder="Tag..."
          type="text"
          onChange={this.handleChange}
          onKeyDown={this.handleKeyPress}
          className={styles.input}
          data-testid="input"
        />
      </div>
    );
  }

  private isTagSelected(tag: ITag) {
    return includes(this.state.selectedTags, tag.id);
  }

  private isInputEmpty(e: React.KeyboardEvent<HTMLInputElement>) {
    return e.currentTarget.value === '';
  }

  private resetInput(e: React.KeyboardEvent<HTMLInputElement>) {
    e.currentTarget.value = '';
  }

  private isValidTagName(name: string) {
    return name.match(/[a-zA-Z]+[ ,]$/g);
  }

  private normalizeTagName(name: string) {
    return name.trim().replace(',', '');
  }
}

export default TagsEditor;
