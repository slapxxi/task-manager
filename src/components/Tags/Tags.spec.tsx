import React from 'react';
import { render } from 'react-testing-library';
import Tags from './Tags';

const tags = [{ id: 'html', name: 'HTML' }, { id: 'css', name: 'CSS' }];

it('renders tags', () => {
  const { container } = render(
    <Tags tags={tags} renderTag={({ tag }) => <div>{tag.name}</div>} />,
  );
  expect(container.firstChild).toMatchSnapshot();
});
