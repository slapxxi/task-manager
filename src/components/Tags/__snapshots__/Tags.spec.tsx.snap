// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders tags 1`] = `
<ul
  class="tags"
  data-testid="tags"
>
  <li
    class="tag"
  >
    <div>
      HTML
    </div>
  </li>
  <li
    class="tag"
  >
    <div>
      CSS
    </div>
  </li>
</ul>
`;
